improved ci image + deploy pipeline

nachog00 opened the pull request:
This PR introduces a new CI workflow setup by creating and configuring several essential files, including `.env.ci-dep-versions`, `Dockerfile.ci`, `ci/build-ci-image.sh`, and `ci/test.sh`. Additionally, it removes the outdated `docker-ci/Dockerfile`.

### Key Changes

1. **Dependency Versions Configuration**: 
   - Added `.env.ci-dep-versions` to manage and document toolchain versions:
     - **Rust**: Version set to 1.87
     - **Go**: Set to the latest version available
     - **Lightwalletd**: Version 0.4.16
     - **Zcashd**: Version 6.2.0
     - **Zebrad**: Version 2.3.0

2. **Dockerfile for CI**:
   - Introduced `Dockerfile.ci` to standardize the CI environment:
     - Multi-stage builds to optimize image size:
       - Use of the official `golang` and `rust` images for building components.
       - Final stage includes essential binaries like `zcashd`, `zcash-cli`, `zebrad`, and `lightwalletd`.
     - Configured for non-root user execution (user: `zaino`, uid: 1000).
     - Integration with `cargo-nextest` for Rust-based testing.

### Considerations and next steps
The .env file might not be the best way to track the versions. I wanted something separated from the Dockerfile and easily loadable.
I intend on exploring `cargo-make` next to better set those values in `Cargo.toml`.

nachog00 committed:
Add new CI Docker setup and remove old one
- Added a new `.env.ci-dep-versions` file to specify dependency versions for CI
- Created `Dockerfile.ci` to build a Docker image with specific versions of zcashd, lightwalletd, and zebrad
- Removed `docker-ci/Dockerfile`, replacing it with a more structured CI setup

Improving CI setup structure by managing dependencies and build processes more efficiently in a new Dockerfile.

In the future, we might want to explore using `cargo make` for scripting the test command, for now, it requires a specific `docker run` command

nachog00 committed:
Update dependencies and streamline Docker setup
- Updated `.env.ci-dep-versions` to change `ZCASH_VERSION` from 6.1.0 to 6.2.0
- Modified `.gitignore` to include `docker_cargo`
- Simplified `Dockerfile.ci` by removing manual build steps for `zcashd` and replaced with a binary pull
- Adjusted `USER` and `GID` ARGs in `Dockerfile.ci` to match the typical user and avoid permission conflicts
- Added Rust toolchain installation in `Dockerfile.ci`
- Introduced new script `ci/build-ci-image.sh` for CI Docker image build
- Added `ci/test.sh` script for running Dockerized tests

This improves maintainability by automating dependency management and simplifying Docker configuration.

nachog00 committed:
Update Docker image in test script
- Changed Docker image tag from `zaino-ci:local` to `zingodevops/zaino-ci:latest` in `ci/test.sh`
- Ensures using the latest CI environment for testing

nachog00 committed:
Add CI/CD workflow for Docker image

- Created `.github/workflows/build-n-push-ci-image.yaml` for Docker image CI/CD
- Workflow is triggered on push to `dev` branch if `.env.ci-dep-versions` file changes
- Extracted version variables from `.env.zaino-versions` to use in Docker build
- Configured Docker image build and push, supporting multi-platform builds and version tagging

nachog00 commented:
Explanation:

 **Dockerfile.ci**:

 - **Multi-stage Builds**:
   - The Dockerfile uses multi-stage builds to compile different components like `lightwalletd`, `zcashd`, and `zebrad`.
   - Each tool is built and retrieved separately to optimize the final image size by only including the necessary binaries.

 - **USER and HOME**:
   - Configuration for a non-root user (`zaino`) ensures better security and compatibility, considering resource permissions.
   - Set up user-specific paths for cargo (`CARGO_HOME`, `CARGO_TARGET_DIR`). These are crucial for managing Rust dependencies 
     and build cache, which the build system and CI tools can leverage.

 **Volume Mounts**:

 When using `ci/test.sh`, `zaino-ci:latest` latest is run with a few volumes mounted:

 - **Persistent Storage**:
   - Volumes like `-v $PWD:/app` map the current working directory to the container's `/app` directory to allow the container 
     to access local files. This is essential for source code retrieval inside the container environment.
   - `-v $PWD/target:/home/zaino/target` persists compiled assets and boost caching 
     and dependency management.

 - **npm Caches**:
   - `-v $PWD/docker_cargo/git:/home/zaino/.cargo/git` and `/docker_cargo/registry` ensure that the Rust package registry 
     cache and Git dependencies are managed persistently, speeding up builds by leveraging previously downloaded resources.

 **GitHub Workflow**:

 - **Environment Management**:
   - Versions for multiple dependencies are stored in `.env.ci-dep-versions`, allowing easy updates and consistent environment 
     replication both locally and in CI/CD.

 - **Automation Steps**:
   - The workflow checks out code, sets up Docker Buildx for multi-platform builds, and authenticates with DockerHub.
   - Dynamic image tagging and version embedding allow more informative image management.

 **Local Script (build-ci-image.sh)**:

 - The build script sources the `.env.ci-dep-versions`, ensuring that local builds match the CI/CD environment precisely.
 - It invokes the Docker build command, injecting versioning parameters directly into the Docker build context, thus mirroring 
   the CI/CD setup locally.

nachog00 committed:
Update .gitignore and add .gitkeep files
- Modified .gitignore to include docker_cargo/**/* for ignoring all files
- Added exception for docker_cargo/**/.gitkeep to retain empty directories
- Created .gitkeep in docker_cargo, docker_cargo/git, and docker_cargo/registry to ensure directories are tracked

`docker_cargo` is the container's .cargo
I tried having it share the host's .cargo, but faced many permission issues.

nachog00 committed:
Switch to root user environment in Dockerfile and test script
- Updated `Dockerfile.ci` to use root user instead of creating a non-root user
- Commented out creation and setup of non-root user in `Dockerfile.ci`
- Modified `USER` and `HOME` arguments to correspond to root
- Altered `ci/test.sh` to use root image for testing purposes

Switching to root user simplifies permissions management and aligns container operations with root-based processes.

PD: this change is temporary, we expect to wind back to rootless when the conflict with self hosted runners is solved.

nachog00 committed:
Update Dockerfile.ci with version arguments and improved run usage
- Added RUST_VERSION, GO_VERSION, LIGHTWALLETD_VERSION, ZCASH_VERSION, and ZEBRA_VERSION arguments to final stage
- Updated LABEL usage instruction for improved run command with volume mounts

Enhance Docker image definition for better version control and usability in CI

zancas commented on .env.ci-dep-versions:
```diff
@@ -0,0 +1,14 @@
+# Rust toolchain version (for rust:<version>-bookworm base image)
+RUST_VERSION=1.87
```
Should we switch to "stable"?

nachog00 replied:
this value is used to template the base image of:
`FROM rust:${RUST_VERSION}-bookworm AS final`
We could go with plain `bookworm` tag and trust their release process ... ?
Although I rather like the idea of consciously tracking this consensus (this .env should act as the source of truth of what we test with) being explicit.

zancas commented on Dockerfile.ci:
```diff
@@ -0,0 +1,88 @@
+# syntax=docker/dockerfile:1.4
+
+#### 🔧 Global Build Arguments and User Config
+ARG RUST_VERSION=1.76.0
```
Why the old Rust version here?

nachog00 replied:
No reason, typo. I'll change it to latest

zancas reviewed (dismissed):
I added a few questions.

nachog00 committed:
Update Rust version in Dockerfile.ci
- Changed RUST_VERSION from 1.76.0 to 1.87.0

nachog00 committed:
Add script for copying and renaming binaries
- Created `ci/copy-bins.sh` to automate binary extraction
- Used Docker to copy binaries from `zaino-ci` image

This is a temporary workaround. Since some tests look for bins in PATH and some in /test_binaries/bins, we have to support both.
Issue is, test_binaries lives in the rootdir which is mounted, hence overwriten.
One solution could be an entrypoint script that copies them over... But we want to move away from that pattern anyways (the one of having bins in test_binaries)

nachog00 committed:
Remove lightwalletd from CI setup
- Removed GO_VERSION and LIGHTWALLETD_VERSION from .env.ci-dep-versions
- Removed ARG for GO_VERSION and LIGHTWALLETD_VERSION from Dockerfile.ci
- Deleted the lightwalletd build stage in Dockerfile.ci
- Removed COPY step for lightwalletd in Dockerfile.ci

Lightwalletd not needed on CI setup anymore

nachog00 committed:
Merge branch 'ci-image-pipeline-root' into ci-image-pipeline-zcashd-registry

nachog00 committed:
Revert "Merge branch 'ci-image-pipeline-root' into ci-image-pipeline-zcashd-registry"

This reverts commit 170218d883f443610e7cb05928aaaa12fb66500a, reversing
changes made to fffcfeb36b22a3cb9d61c5a02a9e34cc68f550c7.

nachog00 committed:
Updated Dockerfile.ci to use root user
- Changed ARG USER and ARG HOME to default to root
- Commented out non-root user setup section

Switched to root user for simplicity and to potentially resolve permission issues during CI builds.
